proposals/nullable-reference-types-specification.md

```diff
@@ -0,0 +1,266 @@
+
+`class?` is a new constraint denoting "possibly nullable reference type", whereas `class` denotes "nonnullable reference type".
+
+The nullability of a type argument or of a constraint does not impact whether the type satisfies the constraint, except where that is already the case today (nullable value types do not satisfy the `struct` constraint). However, if the type argument does not satisfy the nullability requirements of the constraint, a warning may be given.
+
+
+# Null state and null tracking
+
+Every expression in a given source location has a *null state*, which indicated whether it is believed to potentially evaluate to null. The null state is either "not null" or "maybe null". The null state is used to determine whether a warning should be given about null-unsafe conversions and dereferences.
+
+## Null tracking for variables
+
+For certain expressions denoting variables or properties, the null state is tracked between occurrences, based on assignments to them, tests performed on them and the control flow between them. This is similar to how definite assignment is tracked for variables. The tracked expressions are the ones of the following form:
+
+```antlr
+tracked_expression
+    : simple_name
+    | this
+    | base
+    | tracked_expression '.' identifier
+    ;
+```
+
+Where the identifiers denote fields or properties.
+
+***Describe null state transitions similar to definite assignment***
+
+## Null state for expressions
+
+The null state of an expression is derived from its form and type, and from the null state of variables involved in it.
+
+## Literals
+
+The null state of a `null` literal is "maybe null". The null state of a `default` literal that is being converted to a type that is known not to be a nonnullable value type is "maybe null". The null state of any other literal is "not null".
+
+## Simple names
+
+If a `simple_name` is not classified as a value, its null state is "not null". Otherwise it is a tracked expression, and its null state is its tracked null state at this source location.
+
+## Member access
+
+If a `member_access` is not classified as a value, its null state is "not null". Otherwise, if it is a tracked expression, its null state is its tracked null state at this source location. Otherwise, if the nullability of its type is nullable or unknown, the null state is "maybe null". Otherwise the null state is "not null".
+
+## Invocation expressions
+
+If an `invocation_expression` invokes a member that is declared with one or more attributes for special null behavior, the null state is determined by those attributes. Otherwise, if the nullability of the type of the expression is nullable or unknown, the null state is "maybe null". Otherwise the null state is "not null".
+
```

Comment:
will the C# langauge specify those attributes?  Or is that considered an open-ended implementation detail?